### PR TITLE
[SEDONA-31] R interface for Apache Sedona

### PR DESCRIPTION
**Pull request by yitao-li**
Signed-off-by: Yitao Li <yitao@rstudio.com>

## Is this PR related to a proposed Issue?

https://issues.apache.org/jira/browse/SEDONA-31

https://github.com/r-spark/sparklyr.sedona/issues/39

## What changes were proposed in this PR?

This PR will make `sparklyr.sedona` the official R interface for Apache Sedona. See https://github.com/r-spark/sparklyr.sedona#overview for an overview of the current state of `sparklyr.sedona`.

Alternatively, we can make `sparklyr.sedona` a git submodule, but I don't see much advantage of doing that (in particular, `sparklyr.sedona` should be updated accordingly whenever new versions of Sedona libraries are about to be released, and it's better for such commits to happen in sync with other changes in the rest of Apache Sedona repo rather than happening in a separate repository). Also, size-wise, `sparklyr.sedona` code base will never be more than 5MB anyways (as exceeding this limit would have lead to `sparklyr.sedona` being non-publishable on CRAN). Anyhow, suggestions are welcome.

- https://github.com/r-spark/sparklyr.sedona will become part of the apache/incubator-sedona repo
- A github actions workflow was created to run `sparklyr.sedona` with snapshot Sedona libraries built from the PR branch (instead of published ones from Maven central repository)

## How was this patch tested?

It was tested by creating a PR from my own fork of Apache Sedona at the moment: see https://github.com/yitao-li/incubator-sedona/actions/runs

## Did this PR include necessary documentation updates?

- There is currently a summary plus some short tutorials of `sparklyr.sedona` functionalities in the README.md file.
- All exported R functions of `sparklyr.sedona` have roxygen-style documentation, per standard practice recommended by CRAN.

### CODE REVIEW

**Comment by jiayuasu** [comment]
@yitao-li Thank you very much for this great PR. I am currently trying to cut a release for Sedona 1.0.1 in one or two days. There are some structures change may happen inside Sedona codebase due to some copyright issues.

I will give you detailed code review once I cut the release.

File: R/sparklyr.sedona/tests/testthat/test-sdf-interface.R
Comment (yitao-li) [inline R/sparklyr.sedona/tests/testthat/test-sdf-interface.R:43]: @jiayuasu The test case that is commented out here is not working yet, but it's almost working.
I tihnk it will be fixed as soon as we address a null-pointer issue with `Adapter.toDf()` (please see https://github.com/apache/incubator-sedona/pull/525 and let me know what you think).

File: .github/workflows/r.yml
Comment (jiayuasu) [inline .github/workflows/r.yml:22]: Could you please  upgrade the CI to support Spark 3.1.1?
Please see our latest CI example https://github.com/apache/incubator-sedona/blob/master/.github/workflows/java.yml#L43

File: R/sparklyr.sedona/DESCRIPTION
Comment (jiayuasu) [inline R/sparklyr.sedona/DESCRIPTION:5]: Could you change the maintainer name to Apache Sedona? See here: https://github.com/apache/incubator-sedona/blob/master/python/setup.py#L33

File: R/sparklyr.sedona/inst/extdata/arealm-tiny.csv
Comment (jiayuasu) [inline R/sparklyr.sedona/inst/extdata/arealm-tiny.csv:1]: Can you re-use the test data provided in https://github.com/apache/incubator-sedona/tree/master/core/src/test/resources, by referring to their relative path? Currently Sedona has too many test data files. It leads to some maintenance-wise and copyright-wise issues. We plan to reuse and remove some of them in the future. See [SEDONA-34](https://issues.apache.org/jira/browse/SEDONA-34)

File: R/sparklyr.sedona/man/sedona_read_geojson_to_typed_rdd.Rd
Comment (jiayuasu) [inline R/sparklyr.sedona/man/sedona_read_geojson_to_typed_rdd.Rd:1]: For the documentation, please do not include generated docs in the PR. Please include the source code of docs in the PR. Generated R docs will be manually integrated with generated html files (based on incubator-sedona/docs/) and uploaded to https://github.com/apache/incubator-sedona-website/tree/asf-site

I believe the R docs itself could be a separate PR since it requires some structure change or doc change.

**Review by jiayuasu — CHANGES_REQUESTED** [review]

**Comment by jiayuasu** [comment]
@yitao-li Thank you again for this great PR. Please see my comments above. I believe two current TO-DOs for this PR is (1) re-use existing test data in core/test/resources by referring to their relative paths. (2) Put the R docs in a separate PR and we may need to have some discussion on it.

Moreover, once this R module gets merged into Apache Sedona, I would like to invite you to be a PMC member of Apache Sedona (of course, this needs to pass the community voting phase). We already have several great members on board: @Sarwat and I are the initial architect of Sedona. @jinxuan @zongsizhang and @netanel246 contribute to Sedona architecture, Shapefile reader and serializer. @Imbruced is in charge of Sedona Python.

File: R/sparklyr.sedona/DESCRIPTION
Comment (yitao-li) [reply R/sparklyr.sedona/DESCRIPTION:5]: Sure. There is one logistical matter we need to think about though: when submitting this R package to CRAN, there will be an email from CRAN to the maintainer's email address containing a confirmation link, and someone will need to click that link each time. Would dev@sedona.apache.org work for that purpose?

Comment (yitao-li) [reply R/sparklyr.sedona/DESCRIPTION:5]: If it's a mailing list then that should be fine. I imagine we will vote before starting a release of this R package to CRAN anyways.

**Comment by lorenzwalthert** [comment]
@yitao-li great to see this happening. Maybe a good moment to run {styler} through it before merging this? I only noted a few deviations from the style guide though...

File: R/sparklyr.sedona/inst/extdata/arealm-tiny.csv
Comment (yitao-li) [reply R/sparklyr.sedona/inst/extdata/arealm-tiny.csv:1]: Sure. Will do.

Besides test data, there are also small number of other data files that will be part of the R package for illustration purposes. I think I can simply replace those ones with random numbers and avoid anything with copy-right issues.

Because nowadays CRAN requires R packages to have runnable examples, and part of `sparklyr.sedona` involves reading spatial files of various formats, a small number of example data files will need to be packaged into sparklyr.sedona`.

**Review by yitao-li — COMMENTED** [review]
applying suggested changes

File: R/sparklyr.sedona/DESCRIPTION
Comment (jiayuasu) [inline R/sparklyr.sedona/DESCRIPTION:8]: Oh I just realized the dev@s.a.o is open to the public and anyone has access to it. Would you please change this to private@sedona.apache.org? This way, only PMC members can control the publish of this doc. Once you are added to Sedona PMC (after voting), you will have access to private@sedona.apache.org.

**Review by jiayuasu — CHANGES_REQUESTED** [review]
@yitao-li Please my comments above. One last thing, can you add Apache file header to all .R source code? This excludes the source of documentation.

Thank you so much for your help!

**Comment by yitao-li** [comment]
@jiayuasu Sure. Thanks for the code review! :+1:

**Review by jiayuasu — APPROVED** [review]

**Pull request merged**
